File: Algorithm/DataStructure/stackAndqueue/linked_list_stack.py
from Algorithm.DataStructure.list_node import ListNode


class LinkedListStack:
    # 基于链表实现栈
    def __init__(self) -> None:
        self._peek: ListNode | None = None
        self._size: int = 0

    # 获取栈的长度
    def size(self) -> int:
        return self._size

    # 判断栈是否为空
    def is_empty(self) -> bool:
        return self._size == 0

    # 入栈
    def push(self, val: int):
        node = ListNode(val)
        node.next = self._peek
        self._peek = node
        self._size += 1
    
    # 出栈
    def pop(self) -> int:
        num = self.peek()
        self._peek = self._peek.next
        self._size -= 1
        return num

    # 访问栈顶元素
    def peek(self) -> int:
        if self.is_empty():
            raise IndexError("栈为空")
        return self._peek.val

    # 转化为列表
    def to_list(self) -> list[int]:
        arr = []
        node = self._peek
        while node:
            arr.append(node.val)
            node = node.next
        arr.reverse()
        return arr
    

File: Algorithm/DataStructure/stackAndqueue/LinkedListStack.java
package Algorithm.DataStructure;

public class LinkedListStack {
    private ListNode stackPeek;
    private int stkSize = 0;

    public LinkedListStack(){
        stackPeek = null;
    }

    // 栈的长度
    public int size(){
        return this.stkSize;
    }

    // 判断栈是否为空
    public boolean isEmpty(){
        return stkSize == 0;
    }

    // 入栈
    public void push(int num){
        ListNode listNode = new ListNode(num);
        listNode.next = this.stackPeek;
        this.stackPeek = listNode;
        stkSize++;
    }

    // 出栈
    public int pop(){
        int num = this.peek();
        this.stackPeek = this.stackPeek.next;
        stkSize--;
        return num;
    }

    // 访问栈顶元素
    public int peek(){
        if (this.isEmpty()){
            throw new IndexOutOfBoundsException();
        }
        return this.stackPeek.val;
    }

    // 将List转化为Array并返回
    public int[] toArray(){
        ListNode node = stackPeek;
        int[] res = new int[this.size()];
        for (int i = res.length - 1; i >= 0; i--){
            res[i] = node.val;
            node = node.next;
        }
        return res;
    }
}
